Полностью готовы тесты русский и казахский. Полная проверка http запросов.

kruprog opened the pull request:
Применена библиотека GuzzleHttp.
Она же позволяет тестировать отправку http запросов.
Все тесты проходят.

kruprog committed:
created composer config and folders

kruprog committed:
rus and qazaq

russian Declension - api and unit tests  completely done.
quazaq Declension - api and unit tests partially done.

kruprog committed:
Update .gitignore

kruprog committed:
added github actions for tests

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
composer to action

kruprog committed:
composer install

kruprog committed:
file not found fix

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update DeclensionRussianTest.php

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
simplified

kruprog committed:
composer cache experiment

https://github.com/actions/cache/blob/main/examples.md#php---composer

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
Update manual-phpunit-php8.1.yml

kruprog committed:
beauty

kruprog committed:
множественное пустым тегом

kruprog committed:
тест удалил дубликат

kruprog committed:
разделил тесты по папкам

kruprog committed:
Переделаны http запросы

Применена библиотека GuzzleHttp.
Переделана функция вызова запроса.
Тесты дополнены проверками параметров запроса.

kruprog committed:
Тесты переписаны с использованием Guzzle Mock.

Убран mockery.
Убраны ненужные webClientBase и HttpRequest.

kruprog committed:
Казахский тест доделан

kruprog committed:
Update RussianDeclensionTest.php

kruprog committed:
Create manual-phpunit-php8.1-noWorkflowCache.yml

kruprog committed:
Update manual-phpunit-php8.1.yml

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -0,0 +1,496 @@
+
+
+
+        $mock = new MockHandler([
+            new Response(200, [], $return_text)
+        ]);
+
+        
+        $handlerStack = HandlerStack::create($mock);
+        
+        $container = [];
+        $history = Middleware::history($container);
+        // Add the history middleware to the handler stack.
+        $handlerStack->push($history);
+                
+        $webClientMock=new WebClient('https://test.uu','testtoken',10,$handlerStack);
+  
+        $testMorpher=new Morpher\Ws3Client\Morpher($webClientMock);
```
Возможно, этот повторяющийся в каждом тесте код стоит вынести в отдельный вспомогательный метод?

kruprog replied:
Сделано

kruprog replied:
сделано

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -0,0 +1,496 @@
+        //check request parameters, headers, uri
+        $request=$transaction['request'];        
+        $this->assertEquals("GET", $request->getMethod());   
+        $this->assertTrue($request->hasHeader('Accept'));
+        $this->assertEquals(["application/json"], $request->getHeaders()['Accept']);
+        $this->assertTrue($request->hasHeader('Authorization'));
+        $this->assertEquals(["Basic testtoken"], $request->getHeaders()['Authorization']);
```
В каждом тесте, наверное, не стоит проверять авторизацию. Можно сделать, чтобы эта проверка была в одном месте?

kruprog replied:
Можно наверно написать отдельный кейс конкретно на проверку авторизации, а из всех остальных убрать.

kruprog replied:
Сделано

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -0,0 +1,257 @@
+        $this->assertNull($genitive);
+    }
+
+}
```
Еще нужны тесты с разбором ошибок, например, «Не найдено русских слов» ([пример теста](https://github.com/morpher-ru/morpher-ws3-js-client/blob/main/test/unit/russian/client.test.js#L136) - только исключение должно называться не MorpherException, а более конкретно - NoRussianWordsFoundException).

kruprog replied:
сделано

bzaar reviewed (changes requested):
См. комментарии.

kruprog committed:
Тест авторизации и заголовков вынесен в отдельный кейс

kruprog committed:
Повторяющийся код в тестах выделен в отдельную функцию

kruprog committed:
Добавлено исключение MorpherError и тесты

kruprog committed:
убрал повторяющийся код

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -493,4 +485,64 @@ public function testQazaqParse_Personal_Success():void
+
+    public function testParse_ExceptionNoWords2(): void
+    {
+        $this->expectException(MorpherError::class);
```
На каждый код ошибки 4xx должен быть свой класс (RussianWordsNotFound, QazaqWordsNotFound и т.п.), наследующий [InvalidArgumentException](https://www.php.net/manual/en/class.invalidargumentexception.php). Код ошибки в конструктор исключения передавать не нужно, т.к. его типа уже достаточно.

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -493,4 +485,64 @@ public function testQazaqParse_Personal_Success():void
+    {
+        $this->expectException(MorpherError::class);
+        $this->expectExceptionCode(496);
+        //$this->expectExceptionMessage('Не найдено казахских слов.');
```
Сообщение тоже нужно проверить.

kruprog replied:
видимо нет, т.к. тут теперь InvalidServerResponse исключение, его типа достаточно.

bzaar replied:
Здесь как раз должно быть не InvalidServerResponse, а QazaqWordsNotFound с соотвествующим сообщением, которое в тесте нужно проверить.

bzaar commented on src/WebClient.php:
```diff
@@ -42,14 +42,36 @@ public function send_old(HttpRequest $request): string {
-
-		$result = $response->getBody();
-
+		try
```
Если function send_old() не нужна, то убрать ее.

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -3,26 +3,50 @@
 
 final class QazaqDeclensionTest extends TestCase
 {
+    public function testAuthorizarion(): void
```
Опечатка, должно быть: testAuthorization().

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -254,4 +227,65 @@ public function testNullGenitive(): void
+        $declensionResult=$testMorpher->russian->Parse($lemma);
+
+    }
+
```
Нужны тесты (и соответствующие типы исключений) для всех возможных ошибок:

![image](https://user-images.githubusercontent.com/4028922/163156000-221b69fd-8131-4451-bb81-07695cba5775.png)

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -254,4 +227,65 @@ public function testNullGenitive(): void
+        $this->expectException(MorpherError::class);
+
+
+        $parseResults=[]; //если пустое тело сообщения об ошибке
```
Если тело ответа неожиданно пустое, то это уже другая ошибка - неправильный ответ сервера (InvalidServerResponse, _не_ наследует InvalidArgumentException),

bzaar reviewed (changes requested):
См. комментарии:

kruprog committed:
Добавлены типы исключений  и тесты

bzaar commented on src/exceptions/NotSpecifiedParameterS.php:
```diff
@@ -0,0 +1,7 @@
+<?php
+namespace Morpher\Ws3Client;
+
+class NotSpecifiedParameterS extends \InvalidArgumentException
```
Пользователь не обязательно знает, что в HTTP API параметр назвается 's'. Он пользуется функцией parse, где он называется lemma. Предлагаю назвать этот класс EmptyStringException. И пускай в качестве сообщения выдается «Передана пустая строка».

bzaar commented on src/russian/Client.php:
```diff
@@ -0,0 +1,53 @@
+		$this->webClient=$webClient;
+	}
+	
+	public function Parse(string $lemma)
```
Добавить флаги.

bzaar commented on src/exceptions/DeclensionNotSupportedUseSpell.php:
```diff
@@ -0,0 +1,7 @@
+<?php
+namespace Morpher\Ws3Client;
+
+class DeclensionNotSupportedUseSpell extends \InvalidArgumentException
```
Поместить в папку russian/exceptions.

bzaar commented on src/exceptions/InvalidFlags.php:
```diff
@@ -0,0 +1,7 @@
+<?php
+namespace Morpher\Ws3Client;
+
+class InvalidFlags extends \InvalidArgumentException
```
Поместить в папку russian/exceptions.

bzaar reviewed (changes requested):
См. комментарии:

kruprog committed:
Добавил флаги и тест флагов.

kruprog committed:
Переделаны русские исключения

bzaar commented on tests/qazaq/QazaqDeclensionTest.php:
```diff
@@ -529,12 +527,12 @@ public function testParse_InvalidServerResponse(): void
 
         $parseResults=[        'code'=>6,
-        'message'=> 'Не указан обязательный параметр: s.']; //тело сообщения об ошибке содержит информацию
+        'message'=> 'Передана пустая строка.']; //тело сообщения об ошибке содержит информацию
```
А вот тут бы я оставил 'Не указан обязательный параметр: s.', т.к. именно это отвечает сервер.

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -0,0 +1,399 @@
+    }
+
+
+    public function testParse_ExceptionNoWords2(): void
```
Переименовать в testParse_InvalidServerResponse().

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -273,11 +302,30 @@ public function testParse_ExceptionNoWords2(): void
+        $this->expectExceptionMessage('Передана пустая строка.');
+
+        $parseResults=[        'code'=>6,
+        'message'=> 'Передана пустая строка.'];
```
Здесь тоже оставить 'Не указан обязательный параметр: s.'

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -313,7 +361,7 @@ public function testParse_DeclensionNotSupportedUseSpell(): void
+        $this->expectException(Russian\InvalidFlags::class);
         $this->expectExceptionMessage('Указаны неправильные флаги.');
 
         $parseResults=[        'code'=>12,
```
Я бы еще протестировал каждую функцию на ошибки связи (таймаут, ошибка DNS и т.п.). Точнее на то, что исключения, которые генерирует Guzzle, не съедаются нашей библиотекой и пользователь может их перехватить и обработать. Для юнит-теста достаточно выбросить из мока Guzzle любое исключение, не специфичное для нашей библиотеки.

bzaar reviewed (changes requested):
См. комментарии:

kruprog committed:
Добавлены тесты исключений в случае сетевых ошибок

bzaar committed:
Rename EmptyString to InvalidArgumentEmptyString

bzaar committed:
gitignore .idea/

kruprog committed:
Добавлена обработка некотрректного json ответа,  и тест ошибки 500

kruprog committed:
Update RussianDeclensionTest.php

kruprog committed:
Новые ошибки от сервера и один тест

kruprog committed:
Добавлены тесты в казахский

kruprog committed:
убрано ненужное

kruprog committed:
пробую интеграционный тест с секретом

kruprog committed:
Update IntergrationTest-phpunit-php8.1.yml

kruprog committed:
Update IntergrationTest-phpunit-php8.1.yml

kruprog committed:
Update RussianDeclensionIntegrationTest.php

kruprog committed:
Update IntergrationTest-phpunit-php8.1.yml

kruprog committed:
Исправлены опечатки. Разработаны интеграционные тесты для russian/declension

kruprog committed:
Update IntegrationBase.php

kruprog committed:
Отладил интеграционные тесты. Разобрался с base64.

kruprog committed:
Почищен код. Настроен autoload своих файлов. Организовано локальное хранилище секрета в php файле,  он в игноре, в гитхаб не попадёт.

kruprog committed:
timeout test

bzaar commented on src/russian/Client.php:
```diff
@@ -10,6 +10,7 @@
 use Morpher\Ws3Client\Russian\DeclensionResult;
 require_once "exceptions/InvalidFlags.php";
 require_once "exceptions/DeclensionNotSupportedUseSpell.php";
+require_once __DIR__."/../exceptions/InvalidServerResponse.php";
```
Почему некоторые require_once начинаются с __DIR__, а другие - нет? В этом есть какой-то глубинный смысл? Я просто PHP не знаю. Если нет, то желательно оформить все одинаково.

kruprog replied:
В своих проектах никогда dir не ставил, и так всё работало.
Но видимо на некоторых серверах может не работать. Про это много обсуждений на форумах.
Потом я просто убрал большинство require, когда смог настроить работающую автозагрузку через composer.

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -411,4 +411,42 @@ public function testParse_NetworkError2(): void
+        $testMorpher=MorpherTestHelper::createMockMorpherWithException(new \GuzzleHttp\Exception\ServerException(
+            'Error 500', 
+            new \GuzzleHttp\Psr7\Request('GET', 'test'), 
+            new \GuzzleHttp\Psr7\Response(200,[],'Error 500')
```
Почему 200? Наверное, должно быть 500?

bzaar commented on src/qazaq/Client.php:
```diff
@@ -38,7 +38,15 @@ public function Parse(string $lemma)
 		}
 
-		$result = json_decode($result_raw,true);
+		try
+		{
+			$result = json_decode($result_raw,true,512,JSON_THROW_ON_ERROR);
+		}
+		catch (\JsonException $ex)
+		{
+			throw new \Morpher\Ws3Client\InvalidServerResponse("Некорректный JSON ответ от сервера",$result_raw);
+		}
```
Наверное, стоит выделить этот try/catch в отдельный метод - еще не раз понадобится, практически в каждой публичной функции. В русской Parse есть такой же код.

bzaar commented on tests/russian/RussianDeclensionTest.php:
```diff
@@ -375,7 +375,7 @@ public function testParse_InvalidFlags(): void
         $lemma='двадцать';
     
-        $declensionResult=$testMorpher->russian->Parse($lemma);
+        $declensionResult=$testMorpher->russian->Parse($lemma,["AAA","BBB"]);
```
Для флагов нужно завести константы (в C# это enum) - такие же, как для [Gender](https://github.com/kruprog/morpher-ws3-php-client/blob/18e2d3733c5ec195c810ebe42c0fb24bb51dbaae/src/russian/Gender.php#L8).

kruprog replied:
В данном примере тестовые неправильные флаги, для них точно не нужно.
Для флагов которые сервер воспринимает - согласен, нужно.

bzaar reviewed (changes requested):
См. комментарии:

bzaar committed:
Add explicit types

bzaar commented:
Можно сделать структуру папок такой?

- src
- tests
  - unit
    - russian
    - ukrainian
    - qazaq
  - integration
    - russian
    - ukrainian
    - qazaq

kruprog committed:
Много

Структура папок
Добавлены константы флагов
Выделена функция JsonDecode

kruprog committed:
Merge branch 'master' of https://github.com/kruprog/morpher-ws3-php-client

kruprog committed:
hhh

bzaar commented on src/qazaq/Client.php:
```diff
@@ -0,0 +1,48 @@
+		{
+			if ($ex->getCode()==5) throw new QazaqWordsNotFound($ex->getMessage());
+
+			throw $ex;
```
В этом случае должно выбрасываться InvalidServerResponse с сообщением «Неизвестный код ошибки». MorpherError не содержит никакой полезной информации для пользователя.

bzaar commented on src/russian/Client.php:
```diff
@@ -0,0 +1,58 @@
+			if ($morpher_code==5) throw new RussianWordsNotFound($msg);
+			if ($morpher_code==12) throw new InvalidFlags($msg);
+			if ($morpher_code==4) throw new DeclensionNotSupportedUseSpell($msg);
+			throw $ex;
```
В этом случае должно выбрасываться InvalidServerResponse с сообщением «Неизвестный код ошибки». MorpherError не содержит никакой полезной информации для пользователя.

bzaar committed:
Tidy up

bzaar committed:
Typo

bzaar committed:
Outdated comment ('valid length')

bzaar commented on tests/integration/russian/RussianDeclensionIntegration2Test.php:
```diff
@@ -0,0 +1,63 @@
+<?php declare(strict_types=1);
```
Думаю, этот файл стоит поднять на уровень выше, т.к. он проверяет общие для всех функций и языков ошибки.

bzaar replied:
И назвать, например, ServiceLevelErrors.php.

kruprog replied:
Однако, в этом файле все вызовы идут через russian:
$testMorpher->russian->Parse($lemma);

мы можем или делать те же самые тесты для других языков, по сути других эндпоинтов.
или делать их только для russian, но тогда действительно зачем поднимать на уровень выше? он же на своём месте получается.

у нас тут нет отдельных независящих от языка запросов например для авторизации. проверяется именно эндпоинт russian/declension на то как он воспринимает некорректный токен.

bzaar replied:
Возможно, вы правы, и стоит иметь такие тесты на каждый эндпойнт. Просто слишком много тестов получится. Нельзя ли написать такой набор тестов один раз и передавать ему в качестве параметра лямбду с вызовом метода, например, russian->Parse(), qazaq->Parse и т.п.?

bzaar reviewed (changes requested):
См. комментарии:

kruprog committed:
неизвестный код ошибки

kruprog committed:
Много

добавлены тесты с пустым токеном
добавлены интеграционные qazaq/declension
много где заменен orpher\Ws3Client\.... на    \Morpher\Ws3Client\...   во избежание случайных ошибок

kruprog committed:
workwlows renamed

kruprog committed:
Тесты на основе лямбда. Сокращено количество тестовых функций.

bzaar commented on tests/integration/russian/RussianDeclensionNoTokenTest.php:
```diff
@@ -0,0 +1,67 @@
+        return '';//NO TOKEN
+    }
+
+    protected static $webClient;
+    protected static $testMorpher;
```
Для чего они protected? Тем более, что это final class.

kruprog replied:
в тестах это не принципиально , можно как угодно написать, хоть public.
Доступ к ним извне класса не предполагается, можно хоть private хоть protected написать
только я редко пользуюсь private,
protected для большего количества ситуаций годится.

bzaar replied:
Мы пишем код один раз, а читаем много раз. У меня слово protected вызывает мгновенную реакцию: «Кто наследники этого класса и сполько их? Кто еще имеет доступ к этим non-readonly статическим полям и кто их может поменять?

В общем, прошу убрать protected и указать типы.

kruprog replied:
готово

bzaar commented on tests/integration/qazaq/QazaqDeclensionTest.php:
```diff
@@ -0,0 +1,475 @@
+        $this->assertEquals("тест", $declensionResult->Nominative);
+
+        
+   /*
+        $this->assertEquals("тесттің", $declensionResult->Genitive); //+'тестің'
+        $this->assertEquals("тестке", $declensionResult->Dative);//+'теске'
+        $this->assertEquals("тестті", $declensionResult->Accusative);//+'тесті'
+        $this->assertEquals("тесттен", $declensionResult->Ablative);//тестен
+        $this->assertEquals("тестте", $declensionResult->Locative);//тесте
+        $this->assertEquals("тестпен", $declensionResult->Instrumental);//теспен
+
+        $this->assertEquals("тесттер", $declensionResult->Plural->Nominative);//тестер
+        $this->assertEquals("тесттертің", $declensionResult->Plural->Genitive);//тестердің
+        $this->assertEquals("тесттерке", $declensionResult->Plural->Dative);//тестерге
+        $this->assertEquals("тесттерті", $declensionResult->Plural->Accusative);//тестерді
+        $this->assertEquals("тесттертен", $declensionResult->Plural->Ablative);//тестерден
+        $this->assertEquals("тесттерте", $declensionResult->Plural->Locative);//тестерде
+        $this->assertEquals("тесттерпен", $declensionResult->Plural->Instrumental);//тестермен
+
+*/
```
А почему эти assert-ы закомментированы?

kruprog replied:
Причина в том что сервер отдаёт грамматически отличные от этих ответы на слово Тест.
На другое слово которое тестируется в соседней функции - всё сходится.

bzaar replied:
Понятно. Используйте, пожалуйста, варинты, которые выдает настоящий сервис (без одной буквы т).

kruprog replied:
готово

bzaar commented on src/russian/Client.php:
```diff
@@ -0,0 +1,61 @@
+	{
+		if (trim($lemma)=='') throw new \Morpher\Ws3Client\InvalidArgumentEmptyString();
+		
+		$query="s=".rawurlencode($lemma);
+		if (!empty($flags))
+		{
+			$query.="&flags=".implode(',',$flags);
+		}
+		$result_raw="";
+		try{
+			$headers=['Accept'=> 'application/json'];
+			$tokenBase64=$this->webClient->getTokenBase64();
+			if (!empty($tokenBase64))
+			{
+				$headers['Authorization']= 'Basic '.$tokenBase64;
+
+			}
```
Предлагаю добавление заголовков Accept, Authorization, а также кодирование параметров при помощи rawurlencode переместить в метод send класса WebClient, т.к. эти действия нужны для всех эндпойнтов.

kruprog replied:
>Accept, Authorization,
у меня есть неопубликованная версия где это почти так и сделано,
выделена отдельная функция внутри WebClient которая создает эти стандартные заголовки.

с параметрами есть нюанс.
метод send и так может принимать параметры в виде массива и их кодировать в rawurlencode, это встроенная фича gizzle.
Но ваши флаги надо писать через запятую, и эту запятую по видимому нельзя кодировать.
Чтобы это обойти, именно здесь я написал свой rawurlencode. Больше нигде.

bzaar replied:
Запятую можно кодировать: https://ws3.morpher.ru/russian/declension?s=%D0%9B%D1%8E%D0%B1%D0%BE%D0%B2%D1%8C%20%D0%A1%D0%BE%D0%BA%D0%BE%D0%BB%D0%BE%D0%B2%D0%B0&flags=feminine%2Ccommon

kruprog replied:
готово

bzaar reviewed (changes requested):
См. ниже.

bzaar committed:
Tidy up

bzaar committed:
Remove comments //dataset #0

kruprog committed:
Доработан казахский тест

kruprog committed:
мелкие изящества

kruprog committed:
выделена WebClient->getStandartHeaders();

kruprog committed:
WebClient спрятан чтобы пользователь его не касался

bzaar committed:
Remove 'use Morpher\Ws3Client\WebClient;' from tests/

bzaar committed:
Remove comments (//теспен)

bzaar committed:
Tidy up Client.php (russian and qazaq)

bzaar committed:
Tidy up QazaqDeclensionTest.php

bzaar committed:
WebClient: remove getToken() and getTokenBase64()

bzaar committed:
WebClient: remove _url as it's not used

bzaar committed:
WebClient: remove isset() checks and they are already covered by empty()

bzaar committed:
WebClient: tidy up

bzaar committed:
Update package metadata in composer.json

kruprog reviewed (commented):
Комментарии в стиле //dataset #0 
очень помогают при отладке, потому что PHPUNIT именно так сообщает о том на какой строчке случилась ошибка:
Assertion error on dataset #1
